keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
+                # the overhead of converting to a dataset.
+                inputs = [inputs]
+
+        def preprocess(x):
+            return self.preprocessor(
+                x,
+                sequence_length=max_length,
+                return_labels=False,
+                # We do not append `"<|endoftext|>"` by default, to let the
+                # generation follow directly from the prompt. If working with a
+                # find tuned model where `"<|endoftext|>"` is desired to signal
+                # the end of a prompt, it can be appended to the prompt string.
+                add_end_token=False,
+            )
 
-        # Pad ragged to dense tensors.
-        padded_shape = (None, max_length)
-        min_length = tf.reduce_min(prompt.row_lengths())
-        input_mask = tf.ones_like(prompt, tf.bool).to_tensor(shape=padded_shape)
-        prompt = prompt.to_tensor(shape=padded_shape)
+        if self.preprocessor is not None:
+            if isinstance(inputs, tf.data.Dataset):
+                inputs = inputs.map(preprocess, tf.data.AUTOTUNE)
+                inputs = inputs.prefetch(tf.data.AUTOTUNE)
+            else:
+                inputs = [preprocess(batch) for batch in inputs]
 
-        # Run the (possibly compiled) generate function on dense inputs.
         generate_function = self.make_generate_function()
-        output = generate_function(prompt, input_mask, min_length)
-
-        # Truncate to ragged by removing tokens after the first end token.
-        end_token_id = self.preprocessor.tokenizer.end_token_id
-        output = truncate_at_token(output, end_token_id, input_mask)
-
-        # Detokenize.
-        output = self.preprocessor.tokenizer.detokenize(output)
-        return tf.squeeze(output, 0) if input_is_scalar else output
+        outputs = []
+        for batch in inputs:
+            token_ids, mask = batch["token_ids"], batch["padding_mask"]
+            # If `preprocessor` is attached, we can stop after end_token_id.
+            end_token_id = None
+            if self.preprocessor is not None:
+                end_token_id = self.preprocessor.tokenizer.end_token_id
+            # Run the compiled generate function.
+            output = generate_function(token_ids, mask, end_token_id)
+
+            if self.preprocessor is not None:
+                # Truncate to ragged by removing tokens after a new end token.
+                output = truncate_at_token(output, end_token_id, mask)
+                # Strip start token if added.
+                if self.preprocessor.add_start_token:
+                    output = output[:, 1:]
+                # Detokenize.
+                output = self.preprocessor.tokenizer.detokenize(output)
+            outputs.append(output)
+
+        outputs = tf.concat(outputs, axis=0)
+        outputs = tf.squeeze(outputs, 0) if input_is_scalar else outputs
+        if self.preprocessor is not None and not return_tensors:
```

Comment:
Why not just let `return_tensors` control everything? This may be very hard for the user to reason about.

Reply:
If `preprocessing` is none, we would be return a list of list of python ints (and the utility below would not work, it would need some reworks). I think the better, keep it simple option would just be to remove `return_tensors`.

If preprocessing -> string output, if not `tf.Tensor` token ids with the same shape that you passed in. WDYT?